### ISSUE TITLE
will nvidia ever be supported

### ISSUE THREAD

**Issue by crocer211** [Alex313031/ChromiumOS#3]
well im on desktop and my pc cant display to my igpu so i can only use discrete so will it ever support nvidia

**Comment by crocer211** [comment]
nvm chromium os is starting to support nvidia and i can get it to boot just crashes to pc reboot after 1 minute at the chromium os logo

**Comment by Alex313031** [comment]
@crocer211 Cloudready has nouveau. ChromiumOS is *JUST* now beggining the preliminary work. There are no chromebooks out with nvidia gpus, excluding ARM64 chromebooks with nvidia Tegra X series SOC's with embedded gpu that uses a mali driver.

I will continue to work on this as I find new info.

BTW chromiumos now has codecs, kernel 5.4, and some more additional packages, including the nice stock wallpaper that was the default for chromiumos from version ~50 to ~80ish

**Comment by crocer211** [comment]
ok well i cant actually use this till nvidia works and that may be in a few months or years from now and i wont use cloud ready mainly cause of personal reasons and they never plan on adding android support and i know it can be done as fydeos has done it

**Comment by VitalyVaryvdin** [comment]
Super relevant. Got interested in chromiumos and luck so far isn't with me, as I've described here https://github.com/arnoldthebat/chromiumos/issues/360

@Alex313031 I'd be very thankful if you would give me some insights of what I'm missing on my own builds.

**Comment by VitalyVaryvdin** [comment]
Didn't try your build yet actually. Its been stock amd64-generic. 
Been fiddling around with overlays tomorrow, put updated linux-firmware in there and used 5.10 kernel, got it to "semi-boot" at Intel machine at least. Just getting stuck at chromium logo forever so far. Previously it just kept rebooting right after selecting image. 
Not sure how to debug this tbh.

**Comment by Alex313031** [comment]
@Firstly, you must enable frecon for anything above 4.19. Secondly, 5.10 is really not ready yet, I would recommend 5.4 , thirdly to use 5.x, cryptohome will fail and you wont be able to actually use it after boot without USE uprev-4-to-5 cryptohome_userdataauth_interface has-kernelnext.

I would strongly recommend using the make.defaults and package.use from this repo, rather than having to show you all the lines that are added. Take a look in there and modify as needed, but it is modified to make a generic amd64 image with kernel 5.4 that is suited for the most devices. You can do a diff between mine and the default ones to see what I've changed.

**Comment by crocer211** [comment]
what stuff was added from fydeos into this build anyway unless i didnt read the whole thing i dont see what you added from fydeos which is highly unstable on nvidia anyway

**Comment by VitalyVaryvdin** [comment]
You're right. Sounds much more reasonable to base my work off yours :)
Tried your image just now, got artifacts all round my screen and couldn't switch to VT2 to confirm if it boots or not.

**Comment by Alex313031** [comment]
@crocer211 Fydeos hasn't worked on nvidia anytime ive tried it with various nvidia cards. I added google drive ebuild, and some USE flags in the make.defaults and package.use to enable codecs and vaapi from them.
@VitalyVaryvdin The 5.4 builds seem to be kinda flaky on amd. And regarding arnolds, that's impossible, arnolds has no code for nvidia, and has never worked with nvidia in my testing. The general todo is to build the kernel yourself with a kconfig that allows out of tree module loading, then add all the ebuilds for things nouveau needs like RDEP and BDEP, including overwriting some that are already in the chromium source with newer versions, and then adding the nouveau ebuild, and make sure all of this is in the chromeos-bsp file. Its probably more complex than that, and I've tried every ChromiumOS distribution out there, and the only one that works with an nvidia gpu is cloudready, and that is part of their closed source code.

**Comment by crocer211** [comment]
@Alex313031  i actually did get Fydeos to work but the chromium browser kept crashing so it kept crashing the system but i was able to install it on a nvidia card i have no access to a intel gpu and your build boots to the chromium splash and reboots after 30 seconds and with a normal build i never could get it to boot to the splash screen it would either say booting the kernel or be a blank screen it even did this with Fydeos till now the build they have now worked for me but with the problem i stated above and the driver that was loaded was nouveau last i looked when i booted Fydeos btw my gpu is a gtx 1060 the 3gb one not the 6gb one

**Comment by VitalyVaryvdin** [comment]
https://github.com/arnoldthebat/chromiumos/issues/6 thats the issue i've been talking about

**Comment by crocer211** [comment]
never seen these options but then again i havent thoroughly looked throu the source code

**Comment by VitalyVaryvdin** [comment]
I've been able to boot on Iris graphics as well, just by modifying ebuild for mesa-llvmpipe a little bit for inclusion of Iris gallium driver. 
Still no progress with nvidia so far, since I couldn't even switch VTs. Stateful parition is encrypted and I can not see logs it produces. Haven't found a way to disable encryption yet, but there's an option to modiy one of init scripts to not encrypt this partition.

**Comment by crocer211** [comment]
your build i can get into the VT2 but i dont have enough time to see any of the errors as it reboots like 5 seconds after switching

**Comment by Alex313031** [comment]
@VitalyVaryvdin What changes did you make to mesa-llvmpipe?
Also, to use kernel 5.x, are you putting uprev-4-to-5 cryptohome_userdataauth_interface has-kernelnext in your make.defaults?

**Comment by VitalyVaryvdin** [comment]
My first working boot was without any of these use tags you mention. However, I've tried it with and without and didn't find any difference in the behavior. Currently left them in-place but pretty sure they could be deleted.

Nouveau kernel module is included and being loaded as well, however fails to init screen (but fbcon works?). I've been trying to find any info on mesa error codes and haven't found any so far. It might be missing something else, like proper env vars setup, but I don't see any related logs except for `nvc0_screen_create base screen init failed -19`

Here's modified mesa-llvmpipe ebuild https://gist.github.com/VitalyVaryvdin/b79c79d01acbf6d4d3dc3939a2cf369e
Basically modified several places by following newer mesa.ebuild from Gentoo repo

**Comment by Alex313031** [comment]
@VitalyVaryvdin WDYM currently left them in place? Did you base your overlay on mine like we talked about earlier and you're saying you tried without those use flags and it still worked? If so, HOW?? The image will build without those and boot, but you cannot use the OS, even in guest mode, as cryptohome wont launch which is used to encrypt the STATEFUL partition and the $HOME dir including the temporary one made in guest mode. Those flags are needed on non-unibuild boards (like amd64-generic) to build an image with kernel 5.x.x that works fully.

Also, Cloudready is becoming Chrome OS Flex, and so some of the source code is already in the ChromiumOS repo.
Chrome OS Flex beta can be tested right now as per > https://support.google.com/chromeosflex/answer/11541904?hl=en&ref_topic=11551271

The board name/codename for this is Reven, and you can see that mesa-reven was added to //chromiumos/src/third_party/ with stuff like nouveau and gbm and drm-shim and broadcom firmware that cloudready has. I think taking a look at this will be useful in getting nvidia to work.
https://source.chromium.org/chromiumos/chromiumos/codesearch/+/main:src/third_party/mesa-reven/src/nouveau/?q=reven&ss=chromiumos
https://source.chromium.org/chromiumos/chromiumos/codesearch/+/main:src/third_party/mesa-reven/README.rst

**Comment by VitalyVaryvdin** [comment]
No, my build is based off generic amd64. Felt like starting from scratch learning the whole process after your builds didn't boot up on my system either, which gave me more reasons to investigate this a little bit.

Well, can't tell you how, but it worked, I'll re-check that when once I get back to it
And I've been building R100, maybe this is just result of Neverware work merging into codebase, dunno

Seen reven, its ebuild is quite different from llvmpipe one, didn't have time to trace it's usage back then tho. However, mesa-llvmpipe builds third-party\mesa, which also has all that nouveau stuff.

CloudReady images didn't boot on my system. I hope they'll release overlay used in Flex

**Comment by crocer211** [comment]
i tried flex as well but had the same issue as vital

**Comment by Alex313031** [comment]
@VitalyVaryvdin @crocer211 what are you guys talking about "vital"

**Comment by Alex313031** [comment]
@VitalyVaryvdin I saw your post on chromiumos-dev. For some reason, both chromium-dev and chromiumos-dev have made it where anytime I try to post or respond, it says "posted, but pending approval." It says in the about that new users will have their stuff pending approval. But i've been a member of both groups for years, and have successfully made posts and replies in the past. My stuff never gets approved, even after months of waiting, and it makes it where I cant even ask why this would happen. Have you had any issues with this?
And would you be able to ask for me, and maybe include my email addresses so someone like mike frysinger or another moderator or long time user can help me out or remove whatever block might be in place.

**Comment by Alex313031** [comment]
@VitalyVaryvdin Also, I did a diff with ur mesa-llvmpipe file, and created my own (all I did was add i915)
Also, I would highly recommend using (or using lines from that suit your needs) the package.use and make.defaults in the profiles dir in this repo.
Another thing I would highly recommend is using build_packages --board=${BOARD} --nowithdebug --nowithautotest when running build_packages to reduce build time and remove debug constructs.
The only I would think you might want to change is in the make.defaults in the kernel section to select your kernel.
Take a look at both of those files if you want and give feedback, I think they're fairly good so far and I've enabled compiler optimizations in make.defaults like AVX and -O3, same as I use for my Thorium project.

**Comment by VitalyVaryvdin** [comment]
Honestly, I don't understand these groups. They tend to respond to ill formed question where author barely understands a thing, and yet don't respond to well formed and reasonable questions. But sure, whenever I'll mention your issue.

I have to look and compare your overlay and mine, we probably should just merge our overlays, if there's any improvement in mine, as they end up being pretty close. I'm just using stripped and modified copy of generic-amd64, with my kernel profiles

I'm currently looking at [this](https://github.com/crosbuilder/CustomBuilds). Pretty messy, but people reported nouveau working there back then.

**Comment by Alex313031** [comment]
@VitalyVaryvdin Yeah I would like to compare ours. Can you share your whole overlay dir?
And that's cool.
And so does your modified mesa-llvmpipe work with iris graphics?

**Comment by crocer211** [comment]
@Alex313031 so i tested your latest build it gets stuck on the chromium bootsplash now and no longer reboots in the VT2 i run dmesg and there is a bunch of frecon errors about master not being released but nouveau is being used it does load it

**Comment by Alex313031** [comment]
@crocer211 That build is not intended to work with nouveau yet. It just has fixes and adds stuff to mesa-llvmpipe

Also, @VitalyVaryvdin my two emails that I use both in chromium-dev and chromium-os-dev groups are Alex313031@gmail.com and circumcisionmovie@gmail.com

**Comment by crocer211** [comment]
@Alex313031 it seems tho it will work on its own in a future version tho

**Comment by Alex313031** [comment]
@VitalyVaryvdin @crocer211 I'm going to try to make a build with mesa-llvm excluded entirely, and with mesa-reven set to satisfy the mesa dependency. It might just be that easy, to get a build that will work with all the GPU's that cloudready supports.

**Comment by VitalyVaryvdin** [comment]
Sorry for late response. Yeah, my modified mesa-llvmpipe successfully booted and worked on: i7-4400mq, i5-8400, i3-9400 and some old Celeron. Intel iGPUs from Braswell through Iris are supported, in other words

Unfortunately, mesa-reven won't work on my end :(

**Comment by Alex313031** [comment]
@VitalyVaryvdin @crocer211 I got mesa-reven working with full nvidia support!! ChromiumOS image uploading soon. Also updated my overlay. Also got google drive working. Take a look at the make.defaults, package.use, chromeos-bsp-2.ebuild, and the google-drive-fs directory that I just updated.

**Comment by crocer211** [comment]
ok i will test it out when you upload it

**Comment by crocer211** [comment]
@Alex313031 so i tested it i dont know if it was supposed to work on nvidia yet or not but it doesnt work yet but as i said i dont know if it was supposed to work or not

**Comment by Alex313031** [comment]
@crocer211 It is *supposed* to work with nvidia, butttttt. It builds mesa-reven and tells everything that normally depends on mesa-llvmpipe to defer to mesa-reven (which is also built with the llvmpipe USE flag). I think i need to also tell it to use gbm not minigbm, as i noticed that while it did work on a GT8600 and a GTX 750 Ti, it did not work on a GTX 970. What card did you test it on? Were you at least able to get to a VT2?

**Comment by crocer211** [comment]
@Alex313031 i have a gtx 1060 3gb and i can get to VT2 yes

**Comment by Alex313031** [comment]
@crocer211 @VitalyVaryvdin Check out my latest release. I got Widevine working!!! Also, built kernel from scratch and added two more packages. (Look at the release notes it has everything thats changed)
https://github.com/Alex313031/ChromiumOS/releases/tag/V102.0.4964.1

**Comment by crocer211** [comment]
@Alex313031 i tried your newest release im still having the same issues but i usually boot the usb in legacy and i noticed it said wrong efi loader when i did that but got to the chromium boot splash i checked dmesg and i got some errors that i dont remember but these errors werent there when i booted into uefi on the usb

**Comment by Alex313031** [comment]
@VitalyVaryvdin @crocer211 I have undergone the large amount of work to again try to get nvidia working.
So far I have set mesa-reven to be used in all packages that call mesa or mesa-llvmpipe, and set mesa-reven to use libglvnd and gbm.
Then, I have build the kernel 5.10 locally with the reven splitconfig that overlays on top of the normal cros kernel config.
Then, I modified waffle to use gbm.
Then, I used equery to list all packages that depend on minigbm, and changed it to mesa-reven instead, and set !media-libs/minibgm in chromeos-bsp so that minigbm is not built as part of libchromeos.

And finally, modified the ozone flags in chromeos-chrome as per @VitalyVaryvdin suggestion.

All of this to use reven's kernel and reven's mesa, while setting it to build the full gbm and point all packages that rely on minigbm to instead use this full gbm. 
Packages are building sucessfully so far, and hopefully I will end up with an image that will work with alot more devices (I also udated linux-firmware and enabled broadcom

**Comment by crocer211** [comment]
well i dont have broadcom i have a pretty recent gigabyte network card that runs the ax210 chipset but i also use powerline ethernet adapter so im fine with network

**Comment by crocer211** [comment]
@Alex313031 but maybe it will work on newer nvidia cards

**Comment by Alex313031** [comment]
@VitalyVaryvdin @crocer211 I was able to build chromeos-chrome with the flags, and use the reven kernel, and use waffle with gbm, however it would not build without minigbm so minigbm is still in there. The image works like last time, but again I don't have any other nvidia cards to test it on.

**Comment by crocer211** [comment]
@Alex313031 i only have the gtx 1060 and cannot access my igpu cause i dont have a vga cable so maybe i can test it and im also thinking about getting a amd gpu to be honest

**Comment by crocer211** [comment]
do we know if this will work on newer nvidia cards or will we have to test them

**Comment by crocer211** [comment]
@Alex313031 gets stuck on splash screen so i went to VT2 and used dmesg there is a nouveau error, nouveau  drm failed to create kernel channel -22

**Comment by Alex313031** [comment]
I think the whole overlay needs to be migrated to build chrome with drm, remove gbm including minigbm from everything, and build mesa-reven with classic, gallium, dri, drm, and libglvnd

**Comment by crocer211** [comment]
@Alex313031 well as i can only view the boot splash cause i use nvidia it looks nice matches your pfp but like thats all i can see at the moment due to nvidia

**Comment by crocer211** [comment]
@Alex313031 dont know if your working on this but have you figured out why it doesnt work on nvidia even after all youve done

**Comment by Alex313031** [comment]
@crocer211 I havent worked on ThoriumOS in a lil while but its just because Im working on thorium and other stuff. I haven't stopped development and will continue working on it soon. And no im still investigating, but I'm at least closer than I was to getting nvidia to work on cros.

**Comment by crocer211** [comment]
@Alex313031 im assuming that thorium is what your mainly still working on

**Comment by Alex313031** [comment]
@VitalyVaryvdin @crocer211 Started work again. Making a new build at M107 today!

**Comment by crocer211** [comment]
@Alex313031 i dont know if it will work on a 2060 as i just upgraded to a 2060

**Comment by Alex313031** [comment]
@VitalyVaryvdin @crocer211 Can you all try the latest release. It worked on a 750ti and 1080ti, but strangely not on a 970.

**Comment by crocer211** [comment]
as i no longer use a nvidia gpu this no longer pertains to me
